### PR TITLE
bpo-16285: Update urllib quoting to RFC 3986

### PR DESCRIPTION
**Pull request by rtnpro**
This pull request enhances urllib quoting from RFC 2396 to RFC 3986.

This is an effort to merge initial efforts at http://bugs.python.org/file34950/0be3805cade1.diff in http://bugs.python.org/issue16285.

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow these steps to help rectify the issue:

1. If you don't have an account on b.p.o, please [create one](http://bugs.python.org/user?@template=register)
2. Make sure your GitHub username is listed in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/22867185/d3aedcf8-f137-11e6-8469-d5f7861a1073.png)
3. If you have not already done so, please sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/)
4. If you just signed the CLA, please **wait at least one US business day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
5. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

**Review by ncoghlan — CHANGES_REQUESTED** [review]
The change looks basically OK, but will need test updates, and probably docs updates as well.

File: Doc/library/urllib.parse.rst [hunk 1]
```diff
@@ -451,13 +451,16 @@ task isn't already covered by the URL parsing functions above.
...
-   digits, and the characters ``'_.-'`` are never quoted. By default, this
+   digits, and the characters ``'_.-~'`` are never quoted. By default, this
    function is intended for quoting the path section of URL. The optional *safe*
    parameter specifies additional ASCII characters that should not be quoted
    --- its default value is ``'/'``.
 
    *string* may be either a :class:`str` or a :class:`bytes`.
 
+   Python 3.7 updates from using RFC 2396 to RFC 3986 to quote URL strings.
+   Now, "~" is included in the set of reserved characters.
```
Comment (ncoghlan) [inline Doc/library/urllib.parse.rst]: This should use a `.. versionchanged: 3.7` note (there are several examples for that earlier in the file).

File: Lib/test/test_urllib.py [hunk 2]
```diff
@@ -761,7 +761,7 @@ def test_never_quote(self):
         do_not_quote = '' .join(["ABCDEFGHIJKLMNOPQRSTUVWXYZ",
                                  "abcdefghijklmnopqrstuvwxyz",
                                  "0123456789",
-                                 "_.-"])
+                                 "_.-~"])
```
Comment (ncoghlan) [inline Lib/test/test_urllib.py:764]: And this is the test case update I missed in my earlier review :)

File: Doc/library/urllib.parse.rst (hunk 1 above)
Comment (rtnpro) [reply Doc/library/urllib.parse.rst]: :+1:

Comment (rtnpro) [reply Doc/library/urllib.parse.rst]: Fixed!

**Comment by rtnpro** [comment]
@ncoghlan Is there anything else that needs to be updated in this PR?

File: Doc/library/urllib.parse.rst [hunk 3]
```diff
@@ -451,13 +451,17 @@ task isn't already covered by the URL parsing functions above.
...
-   digits, and the characters ``'_.-'`` are never quoted. By default, this
+   digits, and the characters ``'_.-~'`` are never quoted. By default, this
    function is intended for quoting the path section of URL. The optional *safe*
    parameter specifies additional ASCII characters that should not be quoted
    --- its default value is ``'/'``.
 
    *string* may be either a :class:`str` or a :class:`bytes`.
 
+   .. versionchanged:: 3.7
+      Moving from RFC 2396 to RFC 3986 for quoting URL strings. Now, "~" is
```
Comment (ncoghlan) [inline Doc/library/urllib.parse.rst]: We typically write these notes in past tense, so a slight tweak here:

> Moved from RFC ...

And in the 2nd sentence:

> "~" is now included...

File: Lib/urllib/parse.py [hunk 4]
```diff
@@ -736,15 +736,18 @@ def quote(string, safe='/', encoding=None, errors=None):
...
     reserved    = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+" |
-                  "$" | ","
+                  "$" | "," | "~"
 
     Each of these characters is reserved in some component of a URL,
     but not necessarily in all of them.
 
+    Python 3.7 updates from using RFC 2396 to RFC 3986 to quote URL strings.
+    Now, "~" is included in the set of reserved characters.
+
```
Comment (ncoghlan) [inline Lib/urllib/parse.py:750]: Sorry I missed this one earlier: there's no need to have the version change info in the docstring, so the change in the RFC reference and the addition of `~` to the set of reserved characters is sufficient here.

**Comment by ncoghlan** [comment]
@rtnpro Almost! Just a couple of small comments on the docs and docstring updates.

**Comment by rtnpro** [comment]
@ncoghlan ``versionchanged`` documentation fixed by https://github.com/python/cpython/pull/173/commits/6ce8001d9d8e4f25df5a3b03757aea8de1dbbf39

**Review by ncoghlan — APPROVED** [review]

**Comment by ncoghlan** [comment]
Assigning to myself to do the final ACKS/NEWS/What's New updates before merging.

**Comment by rtnpro** [comment]
Squashed commits!

**Comment by ncoghlan** [comment]
Added the sprint label, as this PR was submitted at the PyCon Pune 2017 core development sprint.

**Comment by openandclose** [comment]
I think the document and the comments are wrong here.

What's changed is the inclusion of '~' to **unreserved** characters.
https://tools.ietf.org/html/rfc3986#section-2.3

So the comments here are wrong.
https://github.com/python/cpython/blob/41439a039dd362bbf19e540c8f01a0e8635d4214/Lib/urllib/parse.py#L742-L743
https://github.com/python/cpython/blob/41439a039dd362bbf19e540c8f01a0e8635d4214/Lib/urllib/parse.py#L748-L750

https://github.com/python/cpython/pull/173/commits/41439a039dd362bbf19e540c8f01a0e8635d4214#diff-01126e4d609357e876a7fd7baa62ce18

And the document here is wrong.
https://github.com/python/cpython/blob/41439a039dd362bbf19e540c8f01a0e8635d4214/Doc/library/urllib.parse.rst#L461-L464

https://github.com/python/cpython/commit/41439a039dd362bbf19e540c8f01a0e8635d4214#diff-67a4980d053b561c26794c63eb5ac1de

**Comment by joernhees** [comment]
Being pinged about this in #2568, i can only repeat: Python never actually followed 2396 to `quote`, but did always quote additional characters such as `()` (see the old https://bugs.python.org/issue12910#msg297675). The problem is that this has been in place for so long that people will definitely rely on it.

Nevertheless, stating that quote was updated to RFC 3986 is still somewhat misleading.

File: Lib/urllib/parse.py [hunk 5]
```diff
@@ -736,15 +736,18 @@ def quote(string, safe='/', encoding=None, errors=None):
     Each part of a URL, e.g. the path info, the query, etc., has a
     different set of reserved characters that must be quoted.
 
-    RFC 2396 Uniform Resource Identifiers (URI): Generic Syntax lists
+    RFC 3986 Uniform Resource Identifiers (URI): Generic Syntax lists
     the following reserved characters.
 
     reserved    = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+" |
-                  "$" | ","
+                  "$" | "," | "~"
```
Comment (joernhees) [inline Lib/urllib/parse.py:743]: This is wrong: "~" is in the set of _UN_reserved chars in RFC 3986, please see https://bugs.python.org/issue12910 and its PR https://github.com/python/cpython/pull/2568

File: Lib/urllib/parse.py (hunk 4 above)
Comment (joernhees) [reply Lib/urllib/parse.py:750]: it's also wrong as stated above.

**Pull request merged**